### ISSUE TITLE
SC2031 missing issues with process substitution (`<()`)

### ISSUE THREAD

**Issue by stdedos** [koalaman/shellcheck#3084]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2031
- My shellcheck version (`shellcheck --version` or "online"):
- [ ] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [ ] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

set -Eeuo pipefail

random_fail() {
  if [ $((RANDOM % 100)) -lt 50 ]; then
    return 1
  fi
  return 0
}

random_fail | tee /tmp/a

# I don't want this to fail immediately, but I want the exit code
# ... also I have other pipes I want to guard against (instead of "consuming" the $?)

(random_fail && RC=0 || RC=$?) | tee /tmp/a
echo "RC: $RC"

# But that triggers SC2030/SC2031. There's a suggestion about using <(), so ... might work?

tee /tmp/a < <(random_fail && RC=0 || RC=$?)
echo "RC: $RC"

# Fun fact: It won't
```

#### Here's what shellcheck currently says:

```
[Line 17:](javascript:setPosition(17, 17))
(random_fail && RC=0 || RC=$?) | tee /tmp/a
                ^-- [SC2030](https://www.shellcheck.net/wiki/SC2030) (info): Modification of RC is local (to subshell caused by (..) group).
 
[Line 18:](javascript:setPosition(18, 11))
echo "RC: $RC"
          ^-- [SC2031](https://www.shellcheck.net/wiki/SC2031) (info): RC was modified in a subshell. That change might be lost.
```

#### Here's what I wanted or expected to see:

_+ Also for l22, l23_

**Comment by stdedos** [comment]
_Solution for the ones who are searching_:

```sh
if random_fail | tee /tmp/a ; then
  RC=0
else
  RC=$?
fi
```

**Comment by ggjulio** [comment]
Also if `errexit` is used process substitution should be imho usually avoided.

It the same issue than when using a subshell in a variable declaration:
> Declare and assign separately to avoid masking return values. (shellcheck [SC2155](https://www.shellcheck.net/wiki/SC2155))

IMO it would make sense to create a new rule :

> Don't use process substitution to avoid masking return values. (shellcheck SCxxxx)

![image](https://github.com/user-attachments/assets/81c9b316-32d2-4509-bdb5-55c5172a4b88)
![image](https://github.com/user-attachments/assets/c526cdfa-94d1-44a7-af02-beee34c61946)



- http://mywiki.wooledge.org/BashFAQ/105

**Comment by wileyhy** [comment]
Assuming these:

random_fail() {
  if [ $((RANDOM % 100)) -lt 50 ]; then
    return 1
  fi
  return 0
}
if random_fail | tee /tmp/a ; then
  RC=0
else
  RC=$?
fi

  I believe the test in this if-fi will always return true if `tee` is able
to open the specified file for writing.

  I'm not exactly sure what you want the code to do. Do you want to capture
an exit code in a variable while also writing that same exit code to a file?
  Wiley
